src/compiler/crystal/config.cr

```diff
@@ -34,8 +34,40 @@ module Crystal
       Time.unix(time).to_s("%Y-%m-%d")
     end
 
-    def self.default_target_triple
-      {{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple
+    @@default_target : Crystal::Codegen::Target?
+
+    def self.default_target : Crystal::Codegen::Target
+      @@default_target ||= begin
+        target = Crystal::Codegen::Target.new({{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple)
+
+        if target.linux?
+          # The linux binary runs as well on linux-gnu as linux-musl, but the
+          # default target needs to match the C library used on the current
+          # system.
```

Comment:
That's not exactly true: this only happens *if* crystal was linked statically.

Reply:
Right. I'll state that more precisely.

This code could also be guarded by `flag?(:static)` but it shouldn't make a difference because a non-static binary can only run on one target platform (which should be equivalent to its default target).